### PR TITLE
Raise if negative weights in shortest path algos.

### LINKED ISSUE #525: [joss] Dijsktra functions do not throw error when encountering negative weights
Functions that implement Dijsktra's algorithm should throw an error when encountering negative weights. Dijkstra's algorithm does not work with negative weights. Example:

```
g = rx.PyDiGraph()
g.extend_from_weighted_edge_list([(0,1, 3), (2,0, 2), (1,2, -2), (1,0, 4)])
```

```
print(rx.all_pairs_dijkstra_path_lengths(g, edge_cost_fn=lambda c: c))
```

```
AllPairsPathLengthMapping{1: PathLengthMapping{0: 0, 2: -2}, 2: PathLengthMapping{0: 2, 1: 5}, 0: PathLengthMapping{2: 1, 1: 3}}
```

The distance from 0 to 1 is returned as 3, but in fact it is 0. Thus the function is giving incorrect results without warnings.

There are other algorithms which can handle negative weights when computing graph distances, assuming the graph has no cycles along which the weights add up to negative numbers. But Dijkstra's can't.

### PR DESCRIPTION
**Pull request by georgios-ts**
### Summary
Some shortest path algorithms, like `dijkstra_shortest_paths`,
`dijkstra_shortest_path_lengths`, `all_pairs_dijkstra_path_lengths`,
`all_pairs_dijkstra_shortest_paths`, `astar_shortest_path`,
`k_shortest_path_lengths` give incorrect results if there are negative
edge weights. This commit adds a check that raises if a user-defined weight
function returns a negative or NaN value.

Closes #525.

### CODE REVIEW

**Review by IvanIsCoding — COMMENTED** [review]
We can simplify the tests, Python has a built-in NaN value we can use

File: tests/digraph/test_astar.py [hunk 1]
```diff
@@ -98,3 +99,19 @@ def test_astar_digraph_with_graph_input(self):
             retworkx.digraph_astar_shortest_path(
                 g, 0, lambda x: x, lambda y: 1, lambda z: 0
             )
+
+    def test_astar_with_invalid_weights(self):
+        g = retworkx.PyDAG()
+        a = g.add_node("A")
+        b = g.add_node("B")
+        g.add_edge(a, b, 7)
+        for invalid_weight in [numpy.nan, -1]:
```
[the remark this answers is not in the document]
Comment (georgios-ts) [reply tests/digraph/test_astar.py]: Done in 604a50f

**Review by mtreinish — APPROVED** [review]
This LGTM, thanks for fixing this. I like the enum based approach for dealing with cost functions it's clever. My only concern is that we're fixing this at the python interface level which is sufficient to close #525 and probably enough. But this does mean for the `retworkx-core` user they'll be responsible for implementing a similar `is_valid_weight()` approach in their usage of retworkx-core's dijkstra function. Which I think is fine (mostly because I looked at doing it in retworkx-core and there isn't really a great way to do it there and handle it for Python cleanly) but we should probably document that in the function's docstring. But we can save that for a followup

**Comment by georgios-ts** [comment]
@mtreinish Yeah, I started doing this in retworkx-core but I had the same feeling, it was just getting messy and I abandoned the approach. Docstrings of shortest path algorithms in retworkx-core already mention that edge costs should be non-negative, see https://github.com/Qiskit/retworkx/blob/main/retworkx-core/src/shortest_path/dijkstra.rs#L38.

**Pull request merged**